File: services/api/test/config.test.ts
import assert from "node:assert/strict";
import { test } from "node:test";

import { databasePoolConfig } from "../src/postgres-repository.js";

test("database configuration prefers a complete connection URL", () => {
  assert.deepEqual(databasePoolConfig({ DATABASE_URL: "postgresql://example/database" }), {
    connectionString: "postgresql://example/database",
    max: 10,
  });
});

test("database configuration accepts secret-backed PostgreSQL fields", () => {
  assert.deepEqual(databasePoolConfig({
    PGHOST: "database.internal",
    PGPORT: "5432",
    PGDATABASE: "blooms",
    PGUSER: "blooms_app",
    PGPASSWORD: "generated-secret",
    PGSSLMODE: "require",
  }), {
    host: "database.internal",
    port: 5432,
    database: "blooms",
    user: "blooms_app",
    password: "generated-secret",
    ssl: { rejectUnauthorized: false },
    max: 10,
  });
});

test("database configuration rejects incomplete cloud credentials", () => {
  assert.throws(
    () => databasePoolConfig({ PGHOST: "database.internal" }),
    /PGPORT must be set/,
  );
});

File: worker_tests/config_test.py
from __future__ import annotations

import os
import unittest
from unittest.mock import patch

from generation_worker.__main__ import database_connection_info


class WorkerConfigurationTests(unittest.TestCase):
    def test_database_url_takes_precedence(self) -> None:
        with patch.dict(os.environ, {"DATABASE_URL": "postgresql://example/blooms"}, clear=True):
            self.assertEqual(database_connection_info(), "postgresql://example/blooms")

    def test_secret_backed_postgres_fields_form_connection_info(self) -> None:
        with patch.dict(
            os.environ,
            {
                "PGHOST": "database.internal",
                "PGPORT": "5432",
                "PGDATABASE": "blooms",
                "PGUSER": "blooms_app",
                "PGPASSWORD": "generated secret",
                "PGSSLMODE": "require",
            },
            clear=True,
        ):
            connection_info = database_connection_info()
        self.assertIn("host=database.internal", connection_info)
        self.assertIn("password='generated secret'", connection_info)
        self.assertIn("sslmode=require", connection_info)


if __name__ == "__main__":
    unittest.main()
